`json.Marshaler` implementors not treated unsafe?

From README.md, things treated as unsafe:

> If a type implements the `json.Marshaler` or `encoding.TextMarshaler` interface (e.g. `json.Number`).

For `encoding.TextMarshaler` it seems to hold:
```go
package main

import (
	"encoding/json"
	"fmt"
)

type Foo int

func (s Foo) MarshalText() (text []byte, err error) {
	return []byte("\"foo\""), nil
}

type Bar struct {
	X Foo
}

func main() {
	x := Bar{X: Foo(0)}
	out, _ := json.Marshal(x)
	fmt.Println(string(out))
}
```
...gives
```
foo.go:20:12: Error return value of `encoding/json.Marshal` is not checked: unsafe type `main.Foo` found
```

However, for `json.Marshaler` it does not seem to hold; if `MarshalText` is changed to `MarshalJSON` in the example, thus making it implement `json.Marshaler`, no error is provoked. This is with 0.2.2.

--- a/errchkjson.go
+++ b/errchkjson.go
@@ -154,7 +154,7 @@ func (e *errchkjson) jsonSafe(t types.Type, level int, seenTypes map[types.Type]
 		return nil
 	}
 
-	if types.Implements(t, textMarshalerInterface()) {
+	if types.Implements(t, textMarshalerInterface()) || types.Implements(t, jsonMarshalerInterface()) {
 		return fmt.Errorf("unsafe type `%s` found", t.String())
 	}
 
@@ -247,7 +247,7 @@ func (e *errchkjson) jsonSafe(t types.Type, level int, seenTypes map[types.Type]
 }
 
 func jsonSafeMapKey(t types.Type) error {
-	if types.Implements(t, textMarshalerInterface()) {
+	if types.Implements(t, textMarshalerInterface()) || types.Implements(t, jsonMarshalerInterface()) {
 		return fmt.Errorf("unsafe type `%s` as map key found", t.String())
 	}
 	switch ut := t.Underlying().(type) {
@@ -268,7 +268,7 @@ func jsonSafeMapKey(t types.Type) error {
 	}
 }
 
-// Construct *types.Interface for interface TextMarshaler
+// Construct *types.Interface for interface encoding.TextMarshaler
 //     type TextMarshaler interface {
 //         MarshalText() (text []byte, err error)
 //     }
@@ -281,6 +281,26 @@ func textMarshalerInterface() *types.Interface {
 					types.NewSlice(
 						types.Universe.Lookup("byte").Type())),
 				types.NewVar(token.NoPos, nil, "err", types.Universe.Lookup("error").Type())),
+			false)),
+	}, nil)
+	textMarshalerInterface.Complete()
+
+	return textMarshalerInterface
+}
+
+// Construct *types.Interface for interface json.Marshaler
+//     type Marshaler interface {
+//         MarshalJSON() ([]byte, error)
+//     }
+//
+func jsonMarshalerInterface() *types.Interface {
+	textMarshalerInterface := types.NewInterfaceType([]*types.Func{
+		types.NewFunc(token.NoPos, nil, "MarshalJSON", types.NewSignature(
+			nil, nil, types.NewTuple(
+				types.NewVar(token.NoPos, nil, "",
+					types.NewSlice(
+						types.Universe.Lookup("byte").Type())),
+				types.NewVar(token.NoPos, nil, "", types.Universe.Lookup("error").Type())),
 			false)),
 	}, nil)
 	textMarshalerInterface.Complete()

--- a/testdata/src/nosafe/a.go
+++ b/testdata/src/nosafe/a.go
@@ -10,11 +10,19 @@ import (
 
 type marshalText struct{}
 
-func (mt marshalText) MarshalText() ([]byte, error) {
+func (_ marshalText) MarshalText() ([]byte, error) {
 	return []byte(`mt`), nil
 }
 
 var _ encoding.TextMarshaler = marshalText(struct{}{})
+
+type marshalJSON struct{}
+
+func (_ marshalJSON) MarshalJSON() ([]byte, error) {
+	return []byte(`mj`), nil
+}
+
+var _ json.Marshaler = marshalJSON(struct{}{})
 
 // JSONMarshalSafeTypesWithNoSafe contains a multitude of test cases to marshal different combinations of types to JSON,
 // that are safe, that is, they will never return an error, if these types are marshaled to JSON.
@@ -266,6 +274,8 @@ type (
 		Stringer             fmt.Stringer
 		Mt                   marshalText
 		MapMarshalTextString map[marshalText]string
+		Mj                   marshalJSON
+		MapMarshalJSONString map[marshalJSON]string
 
 		C128         complex128
 		C128Ptr      *complex128
@@ -301,6 +311,8 @@ func JSONMarshalSafeStructWithUnexportedFieldsWithNoSafe() {
 		stringer             fmt.Stringer                    // unsafe unexported
 		mt                   marshalText                     // unsafe unexported
 		mapMarshalTextString map[marshalText]string          // unsafe unexported
+		mj                   marshalJSON                     // unsafe unexported
+		mapMarshalJSONString map[marshalJSON]string          // unsafe unexported
 		unexportedStruct     ExportedUnsafeAndInvalidStruct  // unsafe unexported
 		unexportedStructPtr  *ExportedUnsafeAndInvalidStruct // unsafe unexported
 
@@ -366,6 +378,8 @@ func JSONMarshalSafeStructWithOmittedFieldsWithNoSafe() {
 		Stringer             fmt.Stringer                    `json:"-"` // unsafe exported but omitted
 		Mt                   marshalText                     `json:"-"` // unsafe exported but omitted
 		MapMarshalTextString map[marshalText]string          `json:"-"` // unsafe exported but omitted
+		Mj                   marshalJSON                     `json:"-"` // unsafe exported but omitted
+		MapMarshalJSONString map[marshalJSON]string          `json:"-"` // unsafe exported but omitted
 		ExportedStruct       ExportedUnsafeAndInvalidStruct  `json:"-"` // unsafe exported but omitted
 		ExportedStructPtr    *ExportedUnsafeAndInvalidStruct `json:"-"` // unsafe exported but omitted
 
@@ -505,6 +519,16 @@ func JSONMarshalUnsafeTypes() {
 	var mapMarshalTextString map[marshalText]string
 	_, _ = json.Marshal(mapMarshalTextString)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `nosafe.marshalText` as map key found"
 	_, err = json.Marshal(mapMarshalTextString) // err is checked
+	_ = err
+
+	var mj marshalJSON
+	_, _ = json.Marshal(mj)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `nosafe.marshalJSON` found"
+	_, err = json.Marshal(mj) // err is checked
+	_ = err
+
+	var mapMarshalJSONString map[marshalJSON]string
+	_, _ = json.Marshal(mapMarshalJSONString)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `nosafe.marshalJSON` as map key found"
+	_, err = json.Marshal(mapMarshalJSONString) // err is checked
 	_ = err
 }
 

--- a/testdata/src/standard/a.go
+++ b/testdata/src/standard/a.go
@@ -10,11 +10,19 @@ import (
 
 type marshalText struct{}
 
-func (mt marshalText) MarshalText() ([]byte, error) {
+func (_ marshalText) MarshalText() ([]byte, error) {
 	return []byte(`mt`), nil
 }
 
 var _ encoding.TextMarshaler = marshalText(struct{}{})
+
+type marshalJSON struct{}
+
+func (_ marshalJSON) MarshalJSON() ([]byte, error) {
+	return []byte(`mj`), nil
+}
+
+var _ json.Marshaler = marshalJSON(struct{}{})
 
 // JSONMarshalSafeTypes contains a multitude of test cases to marshal different combinations of types to JSON,
 // that are safe, that is, they will never return an error, if these types are marshaled to JSON.
@@ -266,6 +274,8 @@ type (
 		Stringer             fmt.Stringer
 		Mt                   marshalText
 		MapMarshalTextString map[marshalText]string
+		Mj                   marshalJSON
+		MapMarshalJSONString map[marshalJSON]string
 
 		C128         complex128
 		C128Ptr      *complex128
@@ -301,6 +311,8 @@ func JSONMarshalSafeStructWithUnexportedFields() {
 		stringer             fmt.Stringer                    // unsafe unexported
 		mt                   marshalText                     // unsafe unexported
 		mapMarshalTextString map[marshalText]string          // unsafe unexported
+		mj                   marshalJSON                     // unsafe unexported
+		mapMarshalJSONString map[marshalJSON]string          // unsafe unexported
 		unexportedStruct     ExportedUnsafeAndInvalidStruct  // unsafe unexported
 		unexportedStructPtr  *ExportedUnsafeAndInvalidStruct // unsafe unexported
 
@@ -366,6 +378,8 @@ func JSONMarshalSafeStructWithOmittedFields() {
 		Stringer             fmt.Stringer                    `json:"-"` // unsafe exported but omitted
 		Mt                   marshalText                     `json:"-"` // unsafe exported but omitted
 		MapMarshalTextString map[marshalText]string          `json:"-"` // unsafe exported but omitted
+		Mj                   marshalJSON                     `json:"-"` // unsafe exported but omitted
+		MapMarshalJSONString map[marshalJSON]string          `json:"-"` // unsafe exported but omitted
 		ExportedStruct       ExportedUnsafeAndInvalidStruct  `json:"-"` // unsafe exported but omitted
 		ExportedStructPtr    *ExportedUnsafeAndInvalidStruct `json:"-"` // unsafe exported but omitted
 
@@ -505,6 +519,16 @@ func JSONMarshalUnsafeTypes() {
 	var mapMarshalTextString map[marshalText]string
 	_, _ = json.Marshal(mapMarshalTextString)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `standard.marshalText` as map key found"
 	_, err = json.Marshal(mapMarshalTextString) // err is checked
+	_ = err
+
+	var mj marshalJSON
+	_, _ = json.Marshal(mj)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `standard.marshalJSON` found"
+	_, err = json.Marshal(mj) // err is checked
+	_ = err
+
+	var mapMarshalJSONString map[marshalJSON]string
+	_, _ = json.Marshal(mapMarshalJSONString)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `standard.marshalJSON` as map key found"
+	_, err = json.Marshal(mapMarshalJSONString) // err is checked
 	_ = err
 }
 

